File: elixir/boutique-inventory/lib/boutique_inventory.ex
defmodule BoutiqueInventory do
  def sort_by_price(inventory) do
    inventory
    |> Enum.sort_by(&(&1.price), :asc)
  end

  def with_missing_price(inventory) do
    inventory
    |> Enum.filter(&(&1.price == nil))
  end

  def update_names(inventory, old_word, new_word) do
    inventory
    |> Enum.map(&(update_item_name(&1, old_word, new_word)))
  end

  defp update_item_name(item, old_word, new_word) do
    item
    |> Map.update!(:name, fn name -> String.replace(name, old_word, new_word) end)
  end

  def increase_quantity(item, count) do
    updated_quantities = item.quantity_by_size
    |> Enum.into(Map.new, fn {key, val} -> {key, val + count} end)

    Map.put(item, :quantity_by_size, updated_quantities)
  end

  def total_quantity(item) do
    item.quantity_by_size
    |> Enum.reduce(0, fn {_key, val}, acc -> val + acc end)
  end
end

File: ruby/boutique-inventory/boutique_inventory.rb
class BoutiqueInventory
  CHEAP_THRESHOLD = 30.0

  def initialize(items)
    @items = items
  end

  def item_names
    @items.map { |i| i[:name] }.sort
  end

  def cheap
    @items.filter { |i| i[:price] < CHEAP_THRESHOLD }
  end

  def out_of_stock
    @items.filter { |i| i[:quantity_by_size].empty? }
  end

  def stock_for_item(name)
    @items.find { |i| i[:name] == name }[:quantity_by_size]
  end

  def total_stock
    @items.map { |i| i[:quantity_by_size].reduce(0) { |sum, (_key, value)| sum + value } }.reduce(0, :+)
  end

  private

  attr_reader :items
end
